types.go

```diff
@@ -0,0 +1,449 @@
+// Composite types in JSON are represented using the Object and Array types.
+// Primitive types in JSON are represented using the Literal type.
+// The Value.Pack method serializes the syntax tree as raw output,
+// which is byte-for-byte identical to the input if no transformations
+// were performed on the value.
+//
+// A HuJSON value can be transformed using the Minimize, Standardize,
+// NormalizeNames, and Reformat methods. Each of these methods mutate the value
+// in place. Call the Clone method in order to preserve the original value.
+// The Minimize and Standardize methods coerces HuJSON into valid standard JSON.
+// The Reformat method formats the value; it is similar to `go fmt`,
+// but instead for the HuJSON and standard JSON format.
+//
+//
+// Grammar
+//
+// The changes to the JSON grammar are:
+//
+//	--- grammar.json
+//	+++ grammar.hujson
+//	@@ -1,13 +1,31 @@
+//	 members
+//	 	member
+//	+	member ',' ws
+//	 	member ',' members
+//
+//	 elements
+//	 	element
+//	+	element ',' ws
+//	 	element ',' elements
+//
+//	+comments
+//	+	"*/"
+//	+	comment comments
+//	+
+//	+comment
+//	+	'0000' . '10FFFF'
+//	+
+//	+linecomments
+//	+	'\n'
+//	+	linecomment linecomments
+//	+
+//	+linecomment
+//	+	'0000' . '10FFFF' - '\n'
+//	+
+//	 ws
+//	 	""
+//	+	"/*" comments
+//	+	"//" linecomments
+//	 	'0020' ws
+//	 	'000A' ws
+//	 	'000D' ws
+//
+package hujson
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"math"
+	"strconv"
+	"unicode/utf8"
+)
+
+// Value is an exact syntactic representation of a JSON value.
+// The starting and ending byte offsets are populated when parsing,
+// but are otherwise ignored when packing.
+type Value struct {
+	// BeforeExtra are the comments and whitespace before Value.
+	// This is the extra after the preceding open brace, open bracket,
+	// colon, comma, or start of input.
+	BeforeExtra Extra
+	// StartOffset is the offset of the first byte in Value.
+	StartOffset int
+	// Value is the JSON value itself without surround extra.
+	Value ValueTrimmed
+	// EndOffset is the offset of the next byte after Value.
+	EndOffset int
+	// AfterExtra are the comments and whitespace after Value.
+	// This is the extra before the succeeding colon, comma, or end of input.
+	AfterExtra Extra
+}
+
+// Clone returns a deep copy of the value.
+func (v Value) Clone() Value {
+	v.BeforeExtra = copyBytes(v.BeforeExtra)
+	v.Value = v.Value.clone()
+	v.AfterExtra = copyBytes(v.AfterExtra)
+	return v
+}
+
+// ValueTrimmed is a JSON value without surround whitespace or comments.
+// This is a sum type consisting of Literal, *Object, or *Array.
+type ValueTrimmed interface {
+	// Kind reports the kind of the JSON value.
+	// It is the first byte of the grammar for that JSON value,
+	// with the exception that JSON numbers are represented as a '0'.
+	//
+	//	'n': null
+	//	'f': false
+	//	't': true
+	//	'"': string
+	//	'0': number
+	//	'{': object
+	//	'[': array
+	//
+	Kind() byte
```

Comment:
don't want to use a Type here instead of plain `byte`?

Reply:
I don't feel strongly about it. If we use a named type, then we probably want to declare constants as well:
```go
type Kind byte

const (
    NullKind Kind   = 'n'
    FalseKind Kind  = 'f'
    TrueKind Kind   = 't'
    StringKind Kind = '"'
    NumberKind Kind = '0'
    ObjectKind Kind = '{' // can't be Object otherwise it conflicts with the Object type
    ArrayKind Kind  = '[' // can't be Array otherwise it conflicts with the Array type
)
```

It feels like additional API surface that isn't worth it.

Reply:
You don't need the consts right away. You can just do the type for now and attach the documentation to the type rather than to the thing that returns it.

Reply:
Done.